File: src/core/dbc-model.js
// DBC object model + registry of loaded databases.
//
// Each loaded .dbc file becomes one Cluster. Signals are addressed everywhere
// by their *qualified name* `cluster/message/signal`, so identical CAN IDs or
// signal names in different files never clash.

export class Cluster {
  constructor(name, fileName) {
    this.name = name; // display name (file name without extension)
    this.fileName = fileName;
    this.messages = new Map(); // key: idWithExtBit >>> 0 -> Message
    this.nodes = []; // BU_ node names
    this.parseErrors = []; // { line, text }
    this.comment = '';
  }

  messageById(idWithExt) {
    return this.messages.get(idWithExt >>> 0) ?? null;
  }
}

export class Message {
  constructor({ id, extended, name, dlc, transmitter }) {
    this.id = id; // raw identifier without ext bit
    this.extended = extended;
    this.name = name;
    this.dlc = dlc; // payload length from DBC (bytes)
    this.transmitter = transmitter;
    this.signals = []; // Signal[]
    this.comment = '';
    this.cluster = null; // back-ref, set on registration
  }

  get multiplexor() {
    return this.signals.find((s) => s.muxRole === 'multiplexor') ?? null;
  }

  get qualifiedName() {
    return `${this.cluster?.name ?? '?'}/${this.name}`;
  }
}

export class Signal {
  constructor(props) {
    Object.assign(
      this,
      {
        name: '',
        startBit: 0, // DBC start bit (LSB for Intel, MSB position for Motorola)
        bitLength: 1,
        byteOrder: 1, // 1 = Intel (little-endian), 0 = Motorola (big-endian)
        signed: false,
        isFloat: false, // SIG_VALTYPE_ 1 = float, 2 = double
        isDouble: false,
        factor: 1,
        offset: 0,
        min: 0,
        max: 0,
        unit: '',
        receivers: [],
        comment: '',
        startValue: null, // GenSigStartValue attribute (raw), null if absent
        valueTable: null, // Map<number, string> from VAL_
        muxRole: 'none', // 'none' | 'multiplexor' | 'multiplexed'
        muxValue: null, // selector value when muxRole === 'multiplexed'
      },
      props,
    );
    this.message = null; // back-ref
  }

  get qualifiedName() {
    return `${this.message?.qualifiedName ?? '?/?'}/${this.name}`;
  }

  /** Physical default value derived from raw GenSigStartValue, or null. */
  get defaultValue() {
    if (this.startValue == null) return null;
    return this.startValue * this.factor + this.offset;
  }
}

export class DbcRegistry {
  constructor() {
    this.clusters = []; // Cluster[], in load order
  }

  clusterByName(name) {
    return this.clusters.find((c) => c.name === name) ?? null;
  }

  add(cluster) {
    // De-duplicate display names: file.dbc loaded twice -> "file (2)"
    const base = cluster.name;
    let n = 2;
    while (this.clusterByName(cluster.name)) cluster.name = `${base} (${n++})`;
    for (const msg of cluster.messages.values()) {
      msg.cluster = cluster;
      for (const sig of msg.signals) sig.message = msg;
    }
    this.clusters.push(cluster);
  }

  remove(name) {
    const i = this.clusters.findIndex((c) => c.name === name);
    if (i >= 0) this.clusters.splice(i, 1);
  }

  /** Resolve "cluster/message/signal" -> Signal or null. */
  signalByQualifiedName(qname) {
    const [cName, mName, sName] = qname.split('/');
    const cluster = this.clusterByName(cName);
    if (!cluster) return null;
    for (const msg of cluster.messages.values()) {
      if (msg.name === mName) {
        return msg.signals.find((s) => s.name === sName) ?? null;
      }
    }
    return null;
  }

  /** All signals across all clusters. */
  *allSignals() {
    for (const cluster of this.clusters) {
      for (const msg of cluster.messages.values()) {
        for (const sig of msg.signals) yield sig;
      }
    }
  }
}

File: python/canviewer/core/dbc_model.py
"""DBC object model + registry. Port of the JS dbc-model."""
from __future__ import annotations


class Signal:
    def __init__(self, **kw):
        self.name = kw.get("name", "")
        self.start_bit = kw.get("start_bit", 0)
        self.bit_length = kw.get("bit_length", 1)
        self.byte_order = kw.get("byte_order", 1)   # 1 Intel, 0 Motorola
        self.signed = kw.get("signed", False)
        self.is_float = kw.get("is_float", False)
        self.is_double = kw.get("is_double", False)
        self.factor = kw.get("factor", 1.0)
        self.offset = kw.get("offset", 0.0)
        self.min = kw.get("min", 0.0)
        self.max = kw.get("max", 0.0)
        self.unit = kw.get("unit", "")
        self.receivers = kw.get("receivers", [])
        self.comment = ""
        self.start_value = None            # GenSigStartValue (raw)
        self.value_table = None            # dict[int,str]
        self.mux_role = kw.get("mux_role", "none")  # none|multiplexor|multiplexed
        self.mux_value = kw.get("mux_value", None)
        self.message = None

    @property
    def qualified_name(self) -> str:
        m = self.message.qualified_name if self.message else "?/?"
        return f"{m}/{self.name}"

    @property
    def default_value(self):
        if self.start_value is None:
            return None
        return self.start_value * self.factor + self.offset


class Message:
    def __init__(self, id, extended, name, dlc, transmitter):
        self.id = id
        self.extended = extended
        self.name = name
        self.dlc = dlc
        self.transmitter = transmitter
        self.signals: list[Signal] = []
        self.comment = ""
        self.cluster = None

    @property
    def multiplexor(self):
        for s in self.signals:
            if s.mux_role == "multiplexor":
                return s
        return None

    @property
    def qualified_name(self) -> str:
        return f"{self.cluster.name if self.cluster else '?'}/{self.name}"


class Cluster:
    def __init__(self, name, file_name):
        self.name = name
        self.file_name = file_name
        self.messages: dict[int, Message] = {}  # idWithExtBit -> Message
        self.nodes: list[str] = []
        self.parse_errors: list[dict] = []
        self.comment = ""
        self.source = None  # raw DBC text (for persistence/round-trip)


class DbcRegistry:
    def __init__(self):
        self.clusters: list[Cluster] = []

    def cluster_by_name(self, name):
        for c in self.clusters:
            if c.name == name:
                return c
        return None

    def add(self, cluster: Cluster):
        base = cluster.name
        n = 2
        while self.cluster_by_name(cluster.name):
            cluster.name = f"{base} ({n})"
            n += 1
        for msg in cluster.messages.values():
            msg.cluster = cluster
            for sig in msg.signals:
                sig.message = msg
        self.clusters.append(cluster)

    def remove(self, name):
        self.clusters = [c for c in self.clusters if c.name != name]

    def clear_all(self):
        self.clusters = []

    def signal_by_qualified_name(self, qname):
        parts = qname.split("/")
        if len(parts) != 3:
            return None
        cname, mname, sname = parts
        cluster = self.cluster_by_name(cname)
        if not cluster:
            return None
        for msg in cluster.messages.values():
            if msg.name == mname:
                for s in msg.signals:
                    if s.name == sname:
                        return s
        return None

    def all_signals(self):
        for c in self.clusters:
            for m in c.messages.values():
                yield from m.signals
